conans/client/graph/graph_binaries.py

```diff
@@ -91,6 +91,9 @@ def _evaluate_cache_pkg(self, node, package_layout, pref, metadata, remote, remo
         if not node.binary:
             node.binary = BINARY_CACHE
             metadata = metadata or package_layout.load_metadata()
+            if pref.id not in metadata.packages:
+                self._evaluate_remote_pkg(node, pref, remote, remotes)
```

Comment:
Calling ``_evaluate_remote_pkg`` is a bit confusing. I see the idea of protecting against wrong insertions, but this ``_evaluate_cache_pkg`` should probably not be called if there is not really a package, so maybe try to evaluate his possible condition before calling ``_evaluate_cache_pkg``:
```python
        # Maybe here, to complement the condition of ``package_id_exists``?
        if package_layout.package_id_exists(pref.id):  # Binary already in cache, check for updates
            self._evaluate_cache_pkg(node, package_layout, pref, metadata, remote, remotes, update)
            recipe_hash = None
```

Reply:
yes, I've checked, and it works as well. the function `_evaluate_cache_pkg` does some things before assertion happens (e.g. it tries to update revision from remote), but I guess it's okay to skip this logic if `pref.id` doesn't exist in `metadata`.

Reply:
nope, unfortunately, an early check breaks several tests...